packages/nimble-components/src/table/specs/table-column-custom-sort-field-hld.md

```diff
@@ -0,0 +1,31 @@
+# Custom table column sort order HLD
+
+## Problem Statement
+
+In some cases, a client may have the need to specify a custom sort order for a column. For example, a table displaying SystemLink tags may have a field in their records for the stringified tag but may not want to sort that column that stringified value. Instead, they want to provide their own sort order for tags that takes each tag's data type and value into account.
+
+## Links to Relevant Work Items and Reference Material
+
+[Azdo Feature 2665131 - Custom column sorting](https://dev.azure.com/ni/DevCentral/_workitems/edit/2665131)
+
+## Implementation / Design
+
+Columns that need support for a custom column sort order will extend their API to allow a client to optionally specify an additional field within the data that specifies the sort order of that column. This field is expected to be a numeric value that is the sorted data index of the record. When the table is sorted by this column, the table will use the `basic` sort operation and sort the records by the additionally specified field rather than the default `operandDataRecordFieldName` associated with the column.
```

Comment:
From a client perspective, a more convenient API might be to provide a custom sort function rather than to annotate their data with an extra field. I.e. change [`accessorFn`](https://github.com/ni/nimble/blob/d37ca77cda0729b42f05b4bac79be8ccc98ceb22/packages/nimble-components/src/table/index.ts#L1138) to:
```ts
(record: TableNode<TData>): TableFieldValue => {
    if (typeof column.columnInternals.customSortFunction === 'function') {
        return column.columnInternals.customSortFunction(record.clientRecord);
    }
    const fieldName = column.columnInternals.operandDataRecordFieldName;
    if (typeof fieldName !== 'string') {
        return undefined;
    }
    return record.clientRecord[fieldName];
}
```
The main drawback I can think of is that we'd be making it possible for clients to shoot themselves in the foot with a slow sort function, but I think documentation can mitigate that. Also, if they wanted the same speed as the other approach, clients could always pre-calculate the sort indices, add a new field to their record, and let the function return that field's value.

Are there other reasons not to prefer this API?

Reply:
The approach you're suggesting goes against the guideline we've been following with the table that we should never call into client code when interacting with the table. Therefore, I don't think we'd want to allow the client to implement a custom sort function -- similar to the fact that we didn't allow them to implement a custom format function for string/number/date/duration columns. Additionally, I'm not sure how this would work for Blazor clients.

Reply:
> the table that we should never call into client code when interacting with the table.

Is the concern (other than performance) that the client could do all sorts of things that might not be handled correctly in the middle of an operation, e.g. deleting rows/columns? Is it substantially different than the client registering event handlers?

> I'm not sure how this would work for Blazor clients.

Good point.

Reply:
Since clients are used to the custom sort function approach, it would be good to list that as an alternative so we have something to point to in case they ask why we changed that API.

And we might want to consider noting the difference in the migration guide for SLE (possibly low ROI, but we did get feedback about teams feeling blindsided by other behavior changes).

Reply:
I've added the custom sort function idea as an alternative.

I'm not sure about putting it in the migration guide. This seems very column specific rather than general to the whole `sl-grid` to `sl-table` migration as a whole. But, if we think that would be helpful for clients, I can see where in that document this information might make sense.

Reply:
Regarding the migration guide, I don't feel strongly that it's necessary. There are likely only two teams doing migrations now: ours and the one who requested this feature. Both of us are well aware of the plan, so writing it down in another place isn't critical.